### ISSUE TITLE
crash if animation manager is not available

### ISSUE THREAD

**Issue by wmjordan** [stefankueng/BowPad#117]
BowPad crashed on Windows Server Core 2016 as soon as it was launched. No interface was shown.
Can you guide me how to find out the problem?

**Comment by stefankueng** [comment]
check the windows event log. There should be an entry in there.
And you could try the [procdump ](https://docs.microsoft.com/en-us/sysinternals/downloads/procdump)tool:
procdump -e -w BowPad.exe
then start BP -> and then attach the dump and pdb and exe here so I can analyse the dump.

**Comment by wmjordan** [comment]
The event log said BowPad was crashed with exception code 0xc0000409 and offset 0x000000000026b429.
I tried procdump. But this time BowPad crashed too fast and procdump was not able to catch it.
There was a BowPad crashed dialog this morning but afterwards I tried again, not more crash dialog (but it actually did crash anyway).

**Comment by stefankueng** [comment]
0xc0000409 means STATUS_STACK_BUFFER_OVERRUN

try this:
procdump -e -x path/to/bowpad.exe
this will make procdump start bowpad itself. Maybe then it can catch the crash?

**Comment by wmjordan** [comment]
I used the -x parameter and managed to created a dump.
Here's the output of the procdump:

````
[16:23:56] Exception: E06D7363.?AV_com_error@@
[16:23:56] Unhandled: E06D7363.?AV_com_error@@
[16:23:56] Dump 1 initiated: c:\Apps\BowPad.exe_201205_162356.dmp
[16:23:57] Dump 1 complete: 1 MB written in 0.3 seconds
[16:23:57] Dump count reached.
```

Here's the dump, EXE and the PDB file.
[BowPad.7z.zip](https://github.com/stefankueng/BowPad/files/5646737/BowPad.7z.zip)

**Comment by wmjordan** [comment]
With WinDbg, we can see this stack trace.

![image](https://user-images.githubusercontent.com/12182730/101237966-cf804e80-3717-11eb-91c5-ebe048b817f0.png)

**Comment by stefankueng** [comment]
you can also open the dump in visual studio and do the analyzing there :)

Problem is that the animation manager couldn't get created for some reason, and then when creating the animation variable the pAnimMgr pointer is null.

So the question is why BP can't create the animation manager object...

**Comment by wmjordan** [comment]
Unfortunately I know little about C++ programming.
It could take me months to figure this out :P.

What's the animation used for?
Could we skip this procedure if `pAnimMgr` is null?

**Comment by stefankueng** [comment]
animation is used in a lot of places to get a better UI experience. For example, if you move the mouse pointer over a tab, you can see that the colors change slowly, not immediately. The same applies to many other UI elements in BP.
I guess I should add checks (in a lot of places) to ensure no crash if the animation manager isn't available...

**Comment by wmjordan** [comment]
Perhaps people who develops Server Core determines that on server administrators should not enjoy the luxury of animation.
Is it possible to mock a series of dummy-do-nothing animation manager and animation variable classes if the manager is not available? So we might not be have to make so many checks around.

**Comment by wmjordan** [comment]
On my personal computer, I noticed the beautiful fading effect when I moved the mouse cursor around the tabs. The tabs changed their colors gradually. But on other parts on the interfaces, I did not notice more animations.

**Comment by stefankueng** [comment]
another is the scrollbars for the editor window, the tab bar buttons (top right), the tooltips of the editor, the collapse-indicators in the editor, ...

**Comment by wmjordan** [comment]
Oh, I see.
They are so beautifully natural that I seldom noticed them.

**Comment by wmjordan** [comment]
Thank you for addressing this problem.
I updated and compiled the code.

On the Server Core machine, it could not yet work.
Partial of the interface was shown, but crashed once more.
The event viewer recorded another crash with exception code 0xc0000005 and offset 0x000000000009f416.
And the crash dump is 
[BowPad.7z.zip](https://github.com/stefankueng/BowPad/files/5648146/BowPad.7z.zip)

I also managed to make a screenshot when it crashed, where the error box read _instruction ....FF416 referenced 0x00 in the memory_--`null `reference?
![image](https://user-images.githubusercontent.com/12182730/101269177-cb4f4200-37a6-11eb-9e31-cbdc5a3e0cd3.png)

The places on the stack was:
````C++

HRESULT ICommand::InvalidateUICommand( UINT32 cmdId, UI_INVALIDATIONS flags, const PROPERTYKEY *key )
{
    return g_pFramework->InvalidateUICommand(cmdId, flags, key); // stack frame 0
}

....

void CFileTree::MarkActiveDocument(bool ensureVisible)
{
    InvalidateUICommand(cmdFileTree, UI_INVALIDATIONS_PROPERTY, &UI_PKEY_Enabled); // stack frame 1
    InvalidateUICommand(cmdFileTree, UI_INVALIDATIONS_PROPERTY, &UI_PKEY_BooleanValue);
    bool bRecurseDone = false;

````

**Comment by wmjordan** [comment]
As the Ribbon was missing on the above screenshot, I did a file search in the System32 folder (and its sub-folders) and found that there was no UIRibbon.dll, UIRibbonRes.dll, UIAnimation.dll, nor Windows.UI.dll.

It seems that running BowPad on Server Core is "mission impossible" where too many dlls are missing. Since those libraries are COM components, simply copying those missing DLL files from another machine with desktop experience enabled onto the BowPad directory might not help--I tried already.
